vm/starnix/starnix.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package starnix
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/sys/targets"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	var _ vmimpl.Infoer = (*instance)(nil)
+	vmimpl.Register("starnix", ctor, true)
+}
+
+type Config struct {
+	// Number of VMs to run in parallel (1 by default).
+	Count int `json:"count"`
+	// Additional command line arguments for the QEMU binary.
+	// Location of Fuchsia checkout, for now we need this to run ffx commands.
+	Fuchsia string `json:"fuchsia"`
+}
+
+type Pool struct {
+	count   int
+	env     *vmimpl.Env
+	cfg     *Config
+	version string
+}
+
+type instance struct {
+	fuchsia  string
+	name     string
+	index    int
+	cfg      *Config
+	version  string
+	args     []string
+	debug    bool
+	workdir  string
+	timeouts targets.Timeouts
+	port     int
+	rpipe    io.ReadCloser
+	wpipe    io.WriteCloser
+	adb      *exec.Cmd
+	starnix  *exec.Cmd
+	executor string
+	merger   *vmimpl.OutputMerger
+	diagnose chan bool
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse starnix vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1, 128]", cfg.Count)
+	}
+	if !osutil.IsDir(cfg.Fuchsia) {
+		return nil, fmt.Errorf("%q is not a valid directory", cfg.Fuchsia)
+	}
+
+	pool := &Pool{
+		count: cfg.Count,
+		env:   env,
+		cfg:   cfg,
+	}
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	for i := 0; ; i++ {
+		inst, err := pool.ctor(workdir, index)
+		if err == nil {
+			return inst, nil
+		}
+		return nil, err
+	}
+}
+
+func (pool *Pool) ctor(workdir string, index int) (vmimpl.Instance, error) {
+	inst := &instance{
+		fuchsia:  pool.cfg.Fuchsia,
+		name:     fmt.Sprintf("VM-%v", index),
+		index:    index,
+		cfg:      pool.cfg,
+		debug:    pool.env.Debug,
+		timeouts: pool.env.Timeouts,
+		workdir:  workdir,
+		executor: filepath.Join(workdir, "adb_executor.sh"),
+	}
+	closeInst := inst
+	defer func() {
+		if closeInst != nil {
+			closeInst.Close()
+		}
+	}()
+
+	var err error
+	inst.rpipe, inst.wpipe, err = osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := inst.boot(); err != nil {
+		return nil, err
+	}
+
+	closeInst = nil
+	return inst, nil
+}
+func (inst *instance) adbPids() ([]string, error) {
```

Comment:
Nit: In what cases would this return multiple PIDs?

Reply:
This function was removed because we are no longer waiting on adb to be online. Marking this as "resolved".